Fix RPL use of `acceptable_rank`

simonduq opened the pull request:
In RFC6550 there are three sets of neighbors (see https://tools.ietf.org/html/rfc6550#section-8.2.1):
- **candidate neighbor set:** neighbors reachable with link-local multicast. May become parent, sibling or child;
- **parent set:** nodes above in the DODAG. With the following property: `A node's Rank MUST be greater than all elements of its DODAG parent set.`;
- **preferred parent:** selected one among the parent set, this is the parent currently in use.

Now Contiki's neighbor table `rpl_parents` has a misleading name; it is the candidate neighbor set, not the parent set. We have all sorts of nodes in the table, including with rank higher than us, and including children. That's actually ok, as we want to keep track of the rank and metric of all neighbors, in part as they might one day move to the parent set or as a preferred parent.

This PR removes the filtering out of nodes leading to a rank above `MaxRankIncrease`, which should be done on the parent set, not the candidate neighbor set.

The PR also includes the following fixes:
- adds a sanity check to `acceptable_rank` (needed because `min_rank` is set to infinite at every version update);
- clarifies `rpl_select_dag`, which actually operates on an instance and a dag, not an instance and a parent.

pablocorbalan commented:
Should we then add a short description of the `rpl_parents` table and detail that the table actually represents the `candidate neighbor set` in line [rpl-dag.c#L77](https://github.com/contiki-os/contiki/blob/c1797e0ffefe15ef9b9278b3d7fb968ba4637938/core/net/rpl/rpl-dag.c#L77) ?. We could also change the name of the table itself I guess.

simonduq commented:
Good point, I've added a description. I agree we should ideally also rename the table (and many functions names and countless variables)

pablocorbalan commented:
Thanks for the update! Now, the table is properly explained in the code. The rest of the changes seem good to me, but I am not familiar with the RPL code to give a clear opinion...

joakimeriksson commented on core/net/rpl/rpl-dag.c:
```diff
@@ -691,29 +696,31 @@ rpl_find_parent_any_dag(rpl_instance_t *instance, uip_ipaddr_t *addr)
+   * and (3) updates the current DAG. */
   rpl_parent_t *last_parent;
-  rpl_dag_t *dag, *end, *best_dag;
+  rpl_dag_t *d, *end, *best_dag;
```
Is there a bit more informative name that we could use for d? Single character variables (beyond i,j) is not optimal when it comes to readability.

simonduq replied:
right, maybe we keep `dag` as the loop variable and come up with something else for the function argument

joakimeriksson commented:
Ok, I think the PR looks good - the main difference is that ContikiRPL is not nullifying parents (maybe not the optimal name of that function) when processing "parent events". It is a small change and it clean things up. Maybe just if we could improve the naming of the dag variable "d" (which I think is a bit short) I would be happy with the PR.

simonduq commented:
Thanks. I'll look into what you suggest and get back to you

simonduq commented:
I addressed your comment on naming of `d`.
As for `process_parent_event`, I couldn't agree more, but I feel this should be part of a general cleanup of `rpl-dag.c`, where we need to simplify the logic of e.g. `process_parent_event`, `rpl_select_dag`, `rpl_select_parent` etc. I'm not going to tackle this before https://github.com/contiki-os/contiki/pull/1403 and https://github.com/contiki-os/contiki/pull/1508 are merged.

simonduq commented:
setting a timeout for merging
